src/huggingface_hub/_upload_large_folder.py

```diff
@@ -118,6 +125,66 @@ def upload_large_folder_internal(
     paths_list = [get_local_upload_paths(folder_path, relpath) for relpath in filtered_paths_list]
     logger.info(f"Found {len(paths_list)} candidate files to upload")
 
+    # Validation checks for repository limits
+    logger.info("Running validation checks on files to upload...")
+
+    # Check 1: Total file count
+    if len(paths_list) > MAX_FILES_PER_REPO:
+        logger.warning(
+            f"You are about to upload {len(paths_list):,} files. "
+            f"This exceeds the recommended limit of {MAX_FILES_PER_REPO:,} files per repository.\n"
+            f"Consider:\n"
+            f"  - Splitting your data into multiple repositories\n"
+            f"  - Using fewer, larger files (e.g., tar archives)\n"
+            f"  - See: https://huggingface.co/docs/hub/repositories-recommendations"
+        )
+
+    # Check 2: Files per folder
+    files_per_folder = Counter(str(Path(paths.path_in_repo).parent) for paths in paths_list)
```

Comment:
I _think_ this check doesn't account for all cases. It only checks the number of files in the last subdirectory. I have seen some repos with a structure like this:

```
data/
    bar00001/
        data.json
        image.png
    bar00002/
        data.json
        image.png
    bar00003/
        data.json
        image.png
...
```

In this case, each sub-directory containing files are small (e.g. only 2 files) but the directory containing all sub-directories can be gigantic (e.g. 10k entries). Would be good to account for this use case (which _I think_ it doesn't)